axum-extra/src/multipart_builder.rs

```diff
@@ -0,0 +1,337 @@
+//! Generate forms to use in responses. You're probably looking for [MultipartForm].
+
+use axum::response::{IntoResponse, Response};
+use http::{header, HeaderMap};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+/// The `Content-Transfer-Encoding` setting for a part.
+#[derive(Debug)]
+pub enum TransferEncoding {
+    /// If not specified, encoding defaults to UTF-8
+    Default,
+    /// If transferring raw binary data that is not guaranteed to be valid UTF-8.
+    Binary,
+}
+
+/// Create multipart forms to be used in API responses.
+/// This struct implements [IntoResponse], and so it can be returned from a handler like normal.
+#[derive(Debug)]
+pub struct MultipartForm {
+    parts: Vec<Part>,
+}
+
+impl MultipartForm {
+    /// Construct a new empty multipart form with no parts.
+    pub fn new() -> Self {
+        MultipartForm { parts: Vec::new() }
+    }
+
+    /// Initialize a new multipart form with the provided vector of parts.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use axum_extra::multipart_builder::{MultipartForm, Part};
+    ///
+    /// let parts: Vec<Part> = vec![Part::text("foo", "abc"), Part::text("bar", "def")];
+    /// let form = MultipartForm::with_parts(parts);
+    /// ```
+    pub fn with_parts(parts: Vec<Part>) -> Self {
+        MultipartForm { parts }
+    }
+
+    /// Add a new [Part] to the form
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use axum_extra::multipart_builder::{MultipartForm, Part};
+    ///
+    /// let mut form = MultipartForm::new();
+    /// form
+    ///     .part(Part::text("foo", "abc"))
+    ///     .part(Part::text("other_field_name", "def"))
+    ///     .part(Part::file("file", "file.txt", vec![0x68, 0x68, 0x20, 0x6d, 0x6f, 0x6d]));
+    /// ```
+    pub fn part(&mut self, part: Part) -> &mut Self {
```

Comment:
I think this should be named so it's clearer what it does, e.g. `add_part` or `push_part`.

Personally, I'm not also sure this API is that useful,wouldn't it be better to just have something like `parts_mut` which would return a mutable reference to the parts vector and then the user can just push there directly, but also remove stuff, clear it, or any other operation they might need to take.

On the other hand they can just build the vector beforehand and then call `from_parts` just before returning the response.

Reply:
You're right, it did feel kind of odd, I was trying to create something similar to  Reqwest's implementation (`.part()`) chained calls, but with `.with_parts()` it feels redundant, so I removed it. I can add `parts_mut` if you feel it would prudent.